File: CPP/Program/DPR.cpp
/*
Saya Jasmine Noor Fawzia [2200598] mengerjakan soal LP1 dalam Mata Kuliah DPBO
untuk keberkahan-Nya maka saya tidak melakukan kecurangan seperti yang telah dispesifikasikan
Aamiin
*/

//import library
//#pragma once //header yang dipake sekali jika ada beberapa file
#include <iostream>
#include <string>

//menggunakan standard namespace
using namespace std;

//mendeklarasi class -> untuk C++, huruf pertama tidak perlu kapital, namun akan lebih baik jika kapital
class DPR{
    //private atribut
    private:
    string id;      //deklarasi atribut untuk id
    string nama;    //deklarasi atribut untuk nama
    string bidang;  //deklarasi atribut untuk nama bidang
    string partai;  //deklarasi atribut untuk nama partai

    //public method
    public:
    //constructur (tidak ada return type)
    DPR(){
        //set id, nama, bidang, dan partai untuk default value
        this->id = "";      //untuk id
        this->nama = "";    //untuk nama
        this->bidang = "";  //untuk bidang
        this->partai = "";  //untuk partai

    }

    //constructur lain menggunakan parameter dapat digunakan jika objek rerdapat parameter
    DPR(string id, string nama, string bidang, string partai){
        //"this" akan di set dengan parameter
        this->id = id;          //untuk id
        this->nama = nama;      //untuk nama
        this->bidang = bidang;  //untuk bidang
        this->partai = partai;  //untuk partai
    }

    //get id
    string get_id(){
        return this->id;
    }

    //set id
    void set_id(string id){
        this->id = id;
    }
    
    //get nama
    string get_nama(){
        return this->nama;
    }

    //set nama
    void set_nama(string nama){
        this->nama = nama;
    }

    //get bidang
    string get_bidang(){
        return this->bidang;
    }

    //set bidang
    void set_bidang(string bidang){
        this->bidang = bidang;
    }

    //get partai
    string get_partai(){
        return this->partai;
    }

    //set partai
    void set_partai(string partai){
        this->partai = partai;
    }

    //public method "behaviors"
    //tambah anggota DPR
    void tambahDPR(list<DPR>& llist){
        DPR anggota;//membuat atribut dari class private DPR

        cout << "Masukkan ID Anggota: ";    //menampilkan masukkan id
        cin >> anggota.id;                  //memasukkan id
        cout << "Masukkan Nama Anggota: ";  //menampilkan masukkan nama
        cin >> anggota.nama;                //memasukkan nama
        cout << "Masukkan Nama Bidang: ";   //menampilkan masukkan bidang
        cin >> anggota.bidang;              //memasukkan bidang
        cout << "Masukkan Nama Partai: ";   //menampilkan masukkan partai
        cin >> anggota.partai;              //memasukkan partai

        llist.push_back(anggota);
        cout << "Data Anggota DPR berhasil ditambahkan.\n";//menampilkan bahwa anggota dpr sudah ditambahkan
    }

    //ubah anggota DPR
    void ubahDPR(list<DPR>& llist){
        string idModify;//membuat atribut untuk menginput id yang akan diubah
        cout << "Masukkan ID Anggota DPR yang ingin diubah: "; //menampilkan masukkan id
        cin >> idModify; //memasukkan id

        //membandingkan id yang sudah ada dengan id yang diinput
        auto it = find_if(llist.begin(), llist.end(), [idModify](const DPR& anggota){return anggota.id == idModify;});

        //jika id yang dibandingkan sama
        if(it != llist.end()){
            cout << "Masukkan perubahan data Anggota DPR:\n"; //menampilkan masukkan partai
            cout << "Nama: ";   //menampilkan masukkan nama
            cin >> it->nama;    //memasukkan nama
            cout << "Bidang: "; //menampilkan masukkan bidang
            cin >> it->bidang;  //memasukkan bidang
            cout << "Partai: "; //menampilkan masukkan partai
            cin >> it->partai;  //memasukkan partai
            cout << "Data Anggota DPR berhasil diubah.\n";//menampilkan bahwa anggota dpr sudah ditambahkan
        }
        //jika id yang dibandingkan berbeda
        else{
            cout << "ID Anggota DPR tersebut tidak ditemukan.\n";//menampilkan bahwa id tidak ada
        }
    }

    //hapus anggota DPR
    void hapusDPR(list<DPR>& llist){
        string idDelete;//membuat atribut untuk menginput id yang akan dihapus
        cout << "Masukkan ID Anggota DPR yang ingin dihapus: "; //menampilkan masukkan id
        cin >> idDelete;//memasukkan id

        //membandingkan id yang sudah ada dengan id yang diinput
        auto it = find_if(llist.begin(), llist.end(), [idDelete](const DPR& anggota){return anggota.id == idDelete;});

        //jika id yang dibandingkan sama
        if(it != llist.end()){
            llist.erase(it);//menghapus data
            cout << "Data Anggota DPR berhasil dihapus.\n";//menampilkan bahwa anggota dpr sudah dihapus
        }
        //jika id yang dibandingkan berbeda
        else{
            cout << "ID Anggota DPR tersebut tidak ditemukan.\n";//menampilkan bahwa id tidak ada
        }
    }

    //menampilkan anggota DPR
    void tampilkanData(list<DPR>& llist) {
        int i = 0;//inisialisasi
        if(llist.empty()){
            cout << "Belum ada data Anggota DPR" << endl;
        }
        else{
            cout << "\nData Anggota DPR:\n";//menampilkan data anggota dpr
            //perulangan untuk menampilkan data
            for(list<DPR>::iterator it = llist.begin(); it != llist.end(); it++){
                cout << (i + 1) << ". " << endl << "ID Anggota DPR: " << it->get_id() << endl << "Nama Anggota DPR: " << it->get_nama() << endl << "Nama Bidang Anggota: " << it->get_bidang() << endl << "Nama Partai Anggota: " << it->get_partai() << endl;
                i++;//iterasi
            }
        }
    }

    //destructors
    ~DPR(){

    }
};

File: Java/Program/DPR.java
//deklarasi class -> untuk Java, huruf pertama pada class harus kapital

import java.util.ArrayList;
import java.util.Scanner;

public class DPR{
    //atribut private
    private String id;      //deklarasi atribut id anggota
    private String name;    //deklarasi atribut nama anggota
    private String bidang;  //deklarasi atribut bidang
    private String partai;  //deklarasi atribut partai

    //constructors
    //tidak ada return type
    public DPR(){
        this.id = "";
        this.name = "";
        this.bidang = "";
        this.partai = "";
    }

    //constructor lain yang terdapat parameter
    public DPR(String id, String name, String bidang, String partai){
        this.id = id;
        this.name = name;
        this.bidang = bidang;
        this.partai = partai;
    }

    //get id
    public String getId(){
        return this.id;
    }

    //set id
    public void setId(String id){
        this.id = id;
    }

    //get name
    public String getName(){
        return this.name;
    }

    //set name
    public void setName(String name){
        this.name = name;
    }

    //get bidang
    public String getBidang(){
        return this.bidang;
    }

    //set bidang
    public void setBidang(String bidang){
        this.bidang = bidang;
    }

    //get partai
    public String getPartai(){
        return this.partai;
    }

    //set partai
    public void setPartai(String partai){
        this.partai = partai;
    }

    //public method "behaviors"
    //tambah anggota dpr
    public static void tambah(ArrayList<DPR> list){
        Scanner sc = new Scanner(System.in);
        System.out.print("Masukkan jumlah anggota DPR yang ingin ditambahkan: ");
        int n = 0;
        //exception handling
        try{
            n = sc.nextInt();//masukkan banyaknya data
        }
        catch(Exception e){
            System.out.println("The input is not an integer!");
        }
        sc.nextLine();

        System.out.println("\nSilahkan masukkan data Anggota DPR!");
        //masukkan data
        for (int i = 0; i < n; i++){
            System.out.println("\nData Anggota DPR ke-" + (i + 1));
            System.out.print("Masukkan ID: ");
            String id = sc.nextLine();

            System.out.print("Masukkan Nama: ");
            String name = sc.nextLine();

            System.out.print("Masukkan Bidang: ");
            String bidang = sc.nextLine();

            System.out.print("Masukkan Partai: ");
            String partai = sc.nextLine();

            DPR anggotaBaru = new DPR(id, name, bidang, partai);
            list.add(anggotaBaru);
        }
        System.out.println("\nData Anggota DPR berhasil ditambahkan.");
    }

    //ubah anggota dpr
    public static void ubah(ArrayList<DPR> list){
        //jika data belum ada
        if(list.isEmpty()){
            System.out.println("Belum ada data anggota DPR.");
        }
        else{
            Scanner sc = new Scanner(System.in);
            System.out.print("Masukkan ID Anggota DPR yang ingin diubah: ");
            String idUbah = sc.nextLine();//mencari id
            boolean ditemukan = false;

            //mengubah data
            for(DPR anggota : list){
                if(anggota.getId().equals(idUbah)){
                    System.out.println("\nID Anggota DPR ditemukan. Masukkan data baru!");
                    System.out.print("Masukkan Nama Baru: ");
                    String namaBaru = sc.nextLine();
                    anggota.setName(namaBaru);

                    System.out.print("Masukkan Bidang Baru: ");
                    String bidangBaru = sc.nextLine();
                    anggota.setBidang(bidangBaru);

                    System.out.print("Masukkan Partai Baru: ");
                    String partaiBaru = sc.nextLine();
                    anggota.setPartai(partaiBaru);
                    System.out.println("Data anggota DPR berhasil diubah.");
                    ditemukan = true;
                    break;
                }
            }

            //jika id tidak ditemukan
            if(!ditemukan){
                System.out.println("ID Anggota DPR tidak ditemukan.");
            }
        }
    }

    //hapus anggota dpr
    public static void hapus(ArrayList<DPR> list){
        //jika belum ada data
        if(list.isEmpty()){
            System.out.println("Belum ada data anggota DPR.");
        }
        else{
            Scanner sc = new Scanner(System.in);
            System.out.print("Masukkan ID Anggota DPR yang ingin dihapus: ");
            String idHapus = sc.nextLine();//mencari id
            boolean ditemukan = false;

            //menghapus data
            for(DPR anggota : list){
                if(anggota.getId().equals(idHapus)){
                    list.remove(anggota);
                    System.out.println("ID anggota DPR ditemukan. Data berhasil dihapus.");
                    ditemukan = true;
                    break;
                }
            }

            //jika id tidak ditemukan
            if(!ditemukan){
                System.out.println("ID Anggota DPR tidak ditemukan.");
            }
        }
    }

    //menampilkan data
    public static void tampilkanData(ArrayList<DPR> list){
        System.out.println("\nData Anggota DPR");
        //jika tidak ada data
        if(list.isEmpty()){
            System.out.println("Belum ada data anggota DPR.");
        }else{
            //menentukan lebar maksimum untuk setiap kolom
            int maxIdLength = 10;//panjang minimal untuk judul kolom id
            int maxNamaLength = 12;//panjang minimal untuk judul kolom nama
            int maxBidangLength = 6;//panjang minimal untuk judul kolom bidang
            int maxPartaiLength = 6;//panjang minimal untuk judul kolom partai

            //menentukan lebar maksimum isi kolom
            for(DPR anggota : list){
                maxIdLength = Math.max(maxIdLength, anggota.getId().length());
                maxNamaLength = Math.max(maxNamaLength, anggota.getName().length());
                maxBidangLength = Math.max(maxBidangLength, anggota.getBidang().length());
                maxPartaiLength = Math.max(maxPartaiLength, anggota.getPartai().length());
            }

            //menentukan lebar kolom agar tidak terlalu sempit
            maxIdLength = Math.max(maxIdLength, "ID Anggota".length());
            maxNamaLength = Math.max(maxNamaLength, "Nama Anggota".length());
            maxBidangLength = Math.max(maxBidangLength, "Bidang".length());
            maxPartaiLength = Math.max(maxPartaiLength, "Partai".length());

            //output
            //menentukan lebar judul kolom
            String format = "| %-"+ (maxIdLength) + "s | %-"+ (maxNamaLength) + "s | %-"+ (maxBidangLength) + "s | %-"+ (maxPartaiLength) + "s |%n";
            //menentukan lebar garis pembatas baris
            String line = "+-" + "-".repeat(maxIdLength) + "-+-" + "-".repeat(maxNamaLength) + "-+-" + "-".repeat(maxBidangLength) + "-+-" + "-".repeat(maxPartaiLength) + "-+";
            System.out.println(line);
            System.out.printf(format, "ID Anggota", "Nama Anggota", "Bidang", "Partai");
            String linee = "+=" + "=".repeat(maxIdLength) + "=+=" + "=".repeat(maxNamaLength) + "=+=" + "=".repeat(maxBidangLength) + "=+=" + "=".repeat(maxPartaiLength) + "=+";
            System.out.println(linee);
            //menampilkan data
            for(DPR anggota : list){
                System.out.printf(format, anggota.getId(), anggota.getName(), anggota.getBidang(), anggota.getPartai());
                System.out.println(line);
            }
        }
    }

    //Java tidak ada destructor
}